pkg/kv/kvserver/rafttrace/rafttrace.go

```diff
@@ -0,0 +1,423 @@
+// still be useful.
+type RaftTracer struct {
+	// m is a map of all the currently traced entries for this replica. The
+	// aggregate size of the map across all replicas is equal to or less than
+	// numRegisteredStore unless the setting changes in which case we flush all
+	// entries on the next register call. We add to numRegistered before we
+	// update m, and delete from m before we remove from numRegistered to keep
+	// this invariant. On a setting change we flush all existing traces on the
+	// next call to register.
+	m syncutil.Map[kvpb.RaftIndex, traceValue]
+
+	// numRegisteredStore is the number of currently registered traces for this
+	// store, not this replica. The number of registered will always be less
+	// than the MaxConcurrentRaftTraces setting. If the setting is lowered, we
+	// flush all traces on all replicas.
+	numRegisteredStore *atomic.Int64
+
+	// numRegisteredReplica is the number of currently registered traces for
+	// this replica. The sum(numRegisteredReplica) <= numRegisteredStore. We set
+	// numRegisteredReplica to MaxInt64 when we close the tracer to prevent new
+	// registrations.
+	numRegisteredReplica atomic.Int64
+
+	// ctx is the ambient context for the replica and is used for remote
+	// traces. It contains the replica/range information. On each trace we
+	// additionally append the unique trace/span IDs.
+	ctx context.Context
+	st  *cluster.Settings
+}
+
+// NewRaftTracer creates a new RaftTracer with the given ambient context for the
+// replica.
+func NewRaftTracer(
+	ctx context.Context, st *cluster.Settings, numRegisteredStore *atomic.Int64,
+) *RaftTracer {
+	return &RaftTracer{ctx: ctx, st: st, numRegisteredStore: numRegisteredStore}
+}
+
+// MaxConcurrentRaftTraces is the maximum number of entries that can be traced
+// at any time on this store. Additional traces will be ignored until the number
+// of traces drops below the limit. Having too many active traces can negatively
+// impact performance as we iterate over all of them for some messages.
+//
+// TODO(baptist): Bump the default to a reasonable value like 10 that balances
+// usefulness with performance impact once we have validated the performance
+// impact.
+var MaxConcurrentRaftTraces = settings.RegisterIntSetting(
+	settings.SystemVisible,
+	"kv.raft.max_concurrent_traces",
+	"the maximum number of tracked raft traces",
+	0,
+	settings.NonNegativeInt,
+)
+
+// reserveSpace checks if should register a new trace. If there are too many
+// registered traces it will not register and return false. The soft invariant
+// is that numRegisteredStore <= numAllowed which can be temporarily violated if
+// the number .  This method will return true if we can add one to the number
+// registered for both the store and replica, otherwise it will return false.
+func (r *RaftTracer) reserveSpace() bool {
+	numAllowed := MaxConcurrentRaftTraces.Get(&r.st.SV)
+	numRegisteredReplica := r.numRegisteredReplica.Load()
```

Comment:
As an extra optimization. Could you check what mutexes are held in all stacks that update `r.numRegisteredReplica`? I have a suspicition that all of them might be holding `Replica.raftMu` or `Replica.mu`. If so, we could change the `Load()` to just read the local variable (but still update it atomically, for other readers).

If so, might add some `RaftMuLocked` / `ReplicaLocked` suffixes to methods that always hold a mutex, and optimize some atomics out.

Reply:
I'd rather do this as a follow-up. I added a TODO to look at this. I wanted to look at some of this if we do a Raft retrospective as it was very difficult to know which locks are held whe.